my_enumerable_module.rb

```diff
@@ -1,75 +1,84 @@
-        result = []
-        for i in 0...self.length
-            if yield(self[i])
-                result.push(self[i])
-            end
-        end
-        result
-    end
-    def my_all?
-        result = true
-        for i in 0...self.length
-            result = result && yield(self[i])
-        end
-        result
-    end
-    def my_any?
-        result = true
-        for i in 0...self.length
-            result = result || yield(self[i])
-        end
-        result
-    end
-    def my_none?
-        result = false
-        for i in 0...self.length
-            result = result || yield(self[i])
-        end
-        !result
+  end
+
+  def my_each_with_index(start_index = 0)
+    for idx in 0...length
+      yield(self[idx], idx + start_index)
     end
-    def my_count
-        self.length unless block_given?
+    self
+  end
+
+  def my_select
+    result = []
+    my_each do |item|
+      result.push(item) if yield(item)
     end
-end
+    result
+  end
 
-#TESTS
-puts "\nmy_each"
-[1,2,3].each { |x| puts x }
-[1,2,3].my_each {|x| puts x}
+  def my_all?
+    result = true
+    my_each do |item|
+      result &&= yield(item)
+    end
+    result
+  end
 
-puts "\nmy_each_with_index"
-#sets the start index
-[:foo, :bar, :baz].each.with_index(2) { |value, index| puts "#{index}: #{value}" }
-[:foo, :bar, :baz].my_each_with_index(6) { |value, index| puts "#{index}: #{value}" }
+  def my_any?
+    result = true
+    my_each do |item|
+      result ||= yield(item)
+    end
+    result
+  end
 
-puts "\nmy_select"
-puts ["apple","berry","tomatoe","pinnepple","cocoa"].select { |x| x=="apple"}
-puts ["apple","berry","tomatoe","pinnepple","cocoa"].my_select { |x| x=="apple"}
+  def my_none?
+    result = false
+    my_each do |item|
+      result ||= yield(item)
+    end
+    !result
+  end
 
-puts "\nmy_all"
-puts %w[ant bear cat].all? { |word| word.length >= 3 }
-puts %w[ant bear cat].my_all? { |word| word.length >= 3 }
+  def my_count
+    counter = 0
+    return length unless block_given?
 
-puts "\nmy_any"
-puts %w[ant bear cat].any? { |word| word.length >= 4 } 
-puts %w[ant bear cat].my_any? { |word| word.length >= 4 } 
+    my_each do |item|
+      counter += 1 if yield item
+    end
+    counter
+  end
 
-puts "\nmy_none"
-puts %w{ant bear cat}.none? { |word| word.length == 4 }
-puts %w{ant bear cat}.my_none? { |word| word.length == 4 }
+  def my_map(procedure = nil)
+    result = []
+    if procedure.nil?
+      my_each do |item|
+        result.push(yield(item))
+      end
+    else
+      my_each do |item|
+        result.push(procedure.call(item))
+      end
+    end
+    result
+  end
 
-puts "\nmy_count"
-puts [1,2,4,2].count   
-puts [1,2,4,2].my_count    
\ No newline at end of file
+  def my_inject(initial = nil)
+    if initial.nil?
+      memo = self[0]
+      idx = 1
+    else
+      idx = 0
+      memo = initial
+    end
+    for i in idx...length
+      memo = yield(memo, self[i])
+    end
+    memo
+  end
+end
```

Comment:
* [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_inject(:+) != [1,2,3,4,5].inject(:+)
```